index/scorch/optimize_knn.go

```diff
@@ -0,0 +1,91 @@
+//  Copyright (c) 2023 Couchbase, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 		http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build vectors
+// +build vectors
+
+package scorch
+
+import (
+	"fmt"
+
+	index "github.com/blevesearch/bleve_index_api"
+	segment_api "github.com/blevesearch/scorch_segment_api/v2"
+)
+
+type OptimizeVR struct {
+	snapshot *IndexSnapshot
+
+	// map of the same field --> vrs
+	vrs map[string][]*IndexSnapshotVectorReader
+}
+
+func (o *OptimizeVR) Finish() error {
+	// for each field, get the faiss index --> invoke the zap func.
+	// for each VR, populate postings list and iterators
+	// by passing the obtained FAISS index and getting similar vectors.
+	// defer close index - just once.
+
+	for i, seg := range o.snapshot.segment {
+		// for each field, get the faiss index --> invoke the zap func.
+		for field, vrs := range o.vrs {
+			// for each VR belonging to that field
+			if sv, ok := seg.segment.(segment_api.VectorSegment); ok {
+				// reading just once per field per segment.
+				faissIndex, err := sv.GetVectorIndex(field)
+				if err != nil {
+					return err
+				}
+				defer faissIndex.Close()
```

Comment:
so in the profiles, i saw that the Close() is like a blocking call in the SimilarVectors() API (it now blocks Finish() API) and adds to the latency. here is my reasoning - the close() basically frees the memory allocated on the C heap (since the faissIndex is resident on the C heap memory) and there is no dependency of it to the goruntime environment. 

so i'm thinking if its better to make it asynchronous so that the closing of the index doesn't block the Finish() API. 

please let me know if this makes sense to you all.

Reply:
yep, good point, done it in the latest commit.